### PR TITLE
catch aliasing violations in memcpy() and std.mem.copy()

### PR DESCRIPTION
**Pull request by shawnl**
I ran some of the tests, but I suspect that there is some code that depends on the forward-copy behavior of the current std.mem.copy().

### CODE REVIEW

**Comment by daurnimator** [comment]
> I suspect that there is some code that depends on the forward-copy behavior of the current std.mem.copy().

Indeed; including in the most recent commit! https://github.com/ziglang/zig/commit/1657bead463656a0a38eaf49b60d8c651e5403ae#diff-00eeca1e7c7014373bdc9b309f6fa38fR37

What should a user use instead if they want a forward copy?

**Comment by emekoi** [comment]
isn't catching this what the `noalias` keyword is intended to do, once it's safety checks are implemented?

**Comment by shawnl** [comment]
It can't catch noalias when it doesn't know the lengths.

I have a WIP patch with mem.move

**Comment by shawnl** [comment]
The windows failure is seemingly unrelated...

File: lib/std/mem.zig [hunk 1]
```diff
@@ -246,30 +246,60 @@ pub const Compare = enum {
...
 /// Copy all of source into dest at position 0.
 /// dest.len must be >= source.len.
-/// dest.ptr must be >= src.ptr.
-pub fn copyBackwards(comptime T: type, dest: []T, source: []const T) void {
-    // TODO instead of manually doing this check for the whole array
-    // and turning off runtime safety, the compiler should detect loops like
-    // this and automatically omit safety checks for loops
-    @setRuntimeSafety(false);
+/// For types larger than 1 this should not be used on shared memory without locks.
+/// see llvm.memmove.element.unordered.atomic
```
Comment (andrewrk) [inline lib/std/mem.zig]: The Zig standard library and language does not intrinsically depend on LLVM and documentation should independently describe how things work without referencing LLVM.

Also this comment is not correct. Even for 1 byte types Zig does not guarantee atomicity unless one uses atomic operations.

File: lib/std/mem.zig [hunk 2]
```diff
@@ -246,30 +246,60 @@ pub const Compare = enum {
...
-    // and turning off runtime safety, the compiler should detect loops like
-    // this and automatically omit safety checks for loops
-    @setRuntimeSafety(false);
-    assert(dest.len >= source.len);
-    for (source) |s, i|
-        dest[i] = s;
+    var n: usize = source.len * @sizeOf(@typeOf(source[0]));
+    var usrc = @ptrToInt(source.ptr);
+    var udest = @ptrToInt(dest.ptr);
+    if (usrc < udest) {
```
Comment (andrewrk) [inline lib/std/mem.zig]: This breaks compile-time invocations of this function. See #382.

**Pull request closed without merging**